warp/yul/ToCairoVisitor.py

```diff
@@ -1,114 +1,195 @@
+            merge_imports(self.imports, builtin_to_cairo.required_imports())
+            if self.preamble:
+                return f"""{builtin_to_cairo.preamble}
+{builtin_to_cairo.function_call}
+{builtin_to_cairo.ref_copy}"""
+            else:
+                return f"{builtin_to_cairo.function_call}"
+        return f"{fun_repr}({args_repr})"
 
-    def visit_expression_statement(self, node: ast.ExpressionStatement):
+    def visit_expression_statement(self, node: ast.ExpressionStatement) -> str:
         if isinstance(node.expression, ast.FunctionCall):
-            self.visit(node.expression)
+            return self.visit(node.expression)
         else:
             # see ast.ExpressionStatement docstring
             raise ValueError("What am I going to do with it? Why is it here?..")
 
-    def visit_variable_declaration(self, node: ast.VariableDeclaration):
+    def visit_variable_declaration(self, node: ast.VariableDeclaration) -> str:
+        self.preamble = False
         if node.value is None:
             decls_repr = "\n".join(
                 self.print(ast.VariableDeclaration(variables=[x], value=ast.Literal(0)))
                 for x in node.variables
             )
-            self.repr_stack.append(decls_repr)
-            return
-        vars_repr = ", ".join("local " + self.print(x) for x in node.variables)
+            self.preamble = True
+            return decls_repr
         value_repr = self.print(node.value)
         if isinstance(node.value, ast.FunctionCall):
-            if not node.variables:
-                self.repr_stack.append(value_repr)
+            function_name: str = node.value.function_name.name
+            function_args: str = ", ".join(self.print(x) for x in node.value.arguments)
+            vars_repr = self.generate_ids_typed(node.variables, function_name)
+            if function_name in YUL_BUILTINS_MAP.keys():
+                builtin_to_cairo = YUL_BUILTINS_MAP[function_name](function_args)
+                merge_imports(self.imports, builtin_to_cairo.required_imports())
+                if not node.variables:
+                    self.preamble = True
+                    return builtin_to_cairo.generated_cairo
+                else:
+                    self.preamble = True
+                    return f"""{builtin_to_cairo.preamble}
+let ({vars_repr}) = {builtin_to_cairo.function_call}
+{builtin_to_cairo.ref_copy}"""
             else:
-                self.repr_stack.append(f"let ({vars_repr}) = {value_repr}")
+                if not node.variables:
+                    self.preamble = True
+                    return value_repr
+                else:
+                    self.preamble = True
+                    return f"let ({vars_repr}) = {value_repr}"
         else:
-            self.repr_stack.append(f"{vars_repr} = {value_repr}")
+            vars_repr = self.generate_ids_typed(node.variables)
+            self.preamble = True
```

Comment:
This preamble is set to True in all branches, can't it be set once in a branch above?